### DISCUSSION TITLE
using existing user management system with Zitadel

### DISCUSSION THREAD

**Discussion by k-p-ani** [zitadel/zitadel#5674] [Q&A]
Hello Everyone,
    I want to know whether is it possible to use Zitadel as OIDC and integrate it with our existing user management system (where currently all users and roles exist), such that on user login, request goes to Zitadel, which connects to out user management system for authentication and on successful authentication Zitadel will generate id token along with claims (using roles from user management system) embedded into id token.

Thanks in advance
Ani.

**Comment by mffap** [comment]
Hey @k-p-ani 

Yes ZITADEL supports [identity brokering](https://zitadel.com/docs/guides/integrate/identity-providers/introduction). You should be able to use your existing system, if that supports OIDC/OAuth. You can then configure it as an external identity provider in ZITADEL: 

<img width="1643" alt="Screenshot 2023-04-12 at 18 26 17" src="https://user-images.githubusercontent.com/1366906/231521609-209a232c-0bf4-4b1d-a263-c551ee575ce4.png">

You can create custom claims or pull roles from your user system with [Actions](https://zitadel.com/docs/apis/actions/introduction) in ZITADEL.

**Reply by k-p-ani** [reply]
Hello @mffap,
 Thanks for quick reply. 
So you are saying it is mandatory that current user management system must support OIDC/OAuth protocol. There is no way to integrate legacy 3 tier user management applications using some kind of hook in Zitadel.
 
 Thanks
Ani

**Reply by mffap** [reply]
Not saying that it's mandatory. Just from your initial question I was assuming your legacy system supports some kind of oidc, ldap layer or issues JWT tokens.
What does the old system offer? does it have an api or just a form?
You could use [Actions](https://zitadel.com/docs/concepts/features/actions) to send https requests and execute code pre-authentication.

**Reply by k-p-ani** [reply]
Hello @mffap 
Old system is a go/react based web app, which has api exposed for business functionality and user management (including roles). so we have login page, user get authenticated against credentials stored in database, once successfully authenticated, api returns an ACL json object (see below the format), which web client use to show and hide various option in the web client. Now we are thinking to use Zitadel, such that when user send login request through web app to Zitadel, Zitadel could authenticate user using our login api and on successful authentication, append below ACL / Roles to **id token** generated by Zitadel, then client decode id token to get ACL / Roles from it and decide which options should be made visible / invisible in the web client based on ACL / Roles
{
  "username": "user-id",
  "acl": [
    {
      "URI": "*",
      "Method": [
        "*"
      ]
    }
  ],
  "created": "2023-04-13T07:47:50.648238249Z",
}

**Reply by mffap** [reply]
For authenticating / migration users you can use either a push or pull approach: https://zitadel.com/docs/guides/migrate/introduction#just-in-time-migration
So you can either use a pre-authentication request to your api or you can create and redirect the user from your legacy system.

Sending custom claims in the id token should be no problem with an Action (complement token flow, triggger: pre access token generation). You should make sure that your app's token setting has Auth Token Type set as "JWT". You can also request the ACL json from you API on the fly with an http request and insert that in the token.

We are just publishing a guide on integation with Google Workforce Identity Federation where we tested something very similar (sending "flat roles" and nickname as custom claims in an id token). Maybe that will help as well, you can checkout the PR already.

**Reply by k-p-ani** [reply]
Hello @mffap 
   Thanks for your guidance. I like the idea of retrieving ACL json from our api. Regarding first part of your answer "So you can either use a pre-authentication request", Can you please provide more information on this, how to do it in Zitadel
   
   Thanks
   Ani

**Comment by hanspoo** [comment]
> Hello Everyone, I want to know whether is it possible to use Zitadel as OIDC and integrate it with our existing user management system (where currently all users and roles exist), such that on user login, request goes to Zitadel, which connects to out user management system for authentication and on successful authentication Zitadel will generate id token along with claims (using roles from user management system) embedded into id token.
> 
> Thanks in advance Ani.

Hi, i have the same need here, a legacy user password database, that must keep running, and would like to have a simple way to read the users from there directly with some kind of provider, miigration and/or syncronization solutions aforementioned doesn't solve this use case, i think a very common one.
